canon/src/main/scala/cats/canon/ApplicativeFromMonad.scala

```diff
@@ -0,0 +1,7 @@
+package cats
+package canon
+
+trait ApplicativeFromMonad[F[_]] extends Applicative[F] with FunctorFromApplicative[F] { self: Monad[F] =>
+  def ap[A, B](fa: F[A])(ff: F[A => B]): F[B] =
+    flatMap(ff)(f => flatMap(fa)(a => pure(f(a))))
+}
```

Comment:
So, one question I have is -- why are we structuring these traits using self-types? My natural inclination would have been to define something like:

```scala
trait ApplicativeFromMonad[F[_]] extends Monad[F] with FunctorFromApplicative[F] {
  def ap[A, B](fa: F[A])(ff: F[A => B]): F[B] =
    flatMap(ff)(f => flatMap(fa)(a => pure(f(a))))
}
```

Is there a reason this doesn't work? The only reason I ask is that seen in isolation its purpose might not be totally apparent. In fact, defined this way, we could name them `CanonicalMonad[F]`, `CanonicalApplicative[F]`, and so on.

What d you think? Is there something I'm missing?